### PR TITLE
Fix files structure to support go mod

### PR DESCRIPTION
**Pull request by sheerun**
How this fixes issues from #69 You can test it by creating some small program that uses my fork (`import "github.com/sheerun/tftp/v2"`) and doing:

```
go mod init foobar
go mod tidy
```

When you do the same for your repository, it fails like described in #69 

Please not you need to create a new tag like `v2.2.1`, otherwise this fix won't work.

### CODE REVIEW

**Comment by pin** [comment]
Wow, this is huge change!

Apology for the slow response BTW.

My concerns:

* Merging this we make #59 completely obsolete :( and that is big and important improvement we need to merge too (it is my bad being so slow with that). I believe we have to merge that because in my opinion it makes sense, fixes single port test lock ups and race detector.
* It is massive change in one commit and wipes blame for a lot of code. I might be outdated on the GitHub stuff, but in my experience having informative blame makes debugging / trying to figure out "why" questions is helpful.
* I think it was my mistake to merge v2 module fix recently, we better should revert, keep it incompatible and instead cut fully rolled v3 module instead.

I'm thinking about the following:

1. Rollback recent #67 
2. Rebase and land #58 
3. Consider regarding moving to vN subdirectory for module code
4. Merge your change in smaller pieces so it easier to track changes in future
5. Cut v3 and introduce non-incompatible v3 package in the process of 4

Not sure if that makes sense, just trying to catch up really on golang package stuff, so please share your thoughts.

Clarification on why I think we need 3.0.0 / v3 module: it is just from reading https://go.dev/ref/mod#non-module-compat - it says 


    A repository that migrates to modules after version v2.0.0 is tagged should usually release a new major version.


Also CC @VictorLowther and @qin-nz for their thoughts on the plan above.

Finally, thanks for the massive diff (even while would be happy to see a stack of diffs doing changes step-by-step, but not always convenient, I understand) and, again, huge apology for being neglecting last year and a half.... Life happens :/

**Comment by pin** [comment]
Ah, forgot one item on the plan:

1.5 fix Travis CI so we have real green stuff on diffs, not the current fake one :)

File: packet.go [hunk 1]
```diff
@@ -4,187 +4,171 @@ import (
 	"bytes"
 	"encoding/binary"
 	"fmt"
+	"strings"
 )
 
+type Packet interface {
+	Unpack(data []byte) error
+	Pack() []byte
+}
```
Comment (pin) [inline packet.go:13]: I haven't dig deep yet, but I so far I love the idea of this interface! Is it a python thingy?

**Comment by qin-nz** [comment]
Release a new major version is recommand by https://go.dev/ref/mod#non-module-compat .  It's not necessery,  and make a new tag (such as `v2.3.0`) is.
I agree to make a new major release v3.  
I recommand to use vN branch instead of vN subdirectory because most repos use branch.

**Comment by qin-nz** [comment]
@sheerun I can't figure out why we need this huge change just for support go mod.  Just make a new tag should work.
Can you explain why we need to modifiy so much files?

**Comment by pin** [comment]
> why we need to modifiy so much files?
I believe (and @sheerun might correct me) aside from module-related stuff there is also a refactoring of the existing code.
Have to admit I haven't looked into the details, but indeed current code need to be refactored for sure. For instance, we need well-defined FSM(TFTP finite state machine) defined as a .go file at some point. In that sense, the idea to interface message types might make sense? But still that is orthogonal to packaging and we should consider making it in steps.

Dunno, short term goal is to not to break things for current users of course :)

**Comment by sheerun** [comment]
I think you need root folder to preserve old codebase (i.e. backward compatibility), and v2 folder for v2 version

Either way this is the only way it worked for me. I've just took some pre v2 tag, put it in root folder, then took current codebase and copied it into v2 folder. This way v2 imports work and go mod doesn't complain like in https://github.com/pin/tftp/pull/70

**Comment by VictorLowther** [comment]
I would not take this as is. It should be split into at least two smaller pull requests - one for the go.mod fixups, and one for all the rest of the changes (which at first glance would requirea new major revision, since they appear to be fairly invasive).

I also prefer branches to subfolders for git major version management.

**Comment by pin** [comment]
@VictorLowther, do you mind rebasing #59? I've tried it yesterday and I believe it fixes tests lock-up and connections leak. Basically you need to rebase against your's previous merge, nobody touched connection.go and sender.go since. So it was pretty copy-paste rebase for me unless I've missed anything. But tests kept passing and stopped handing up at least :)

**Comment by pin** [comment]
Just to clarify - my current plan:

1. Revert recent module v2 stuff - let's keep 2.x.x incompatible as go instructions suggest.
2. Merge #59
3. Fix Travis (It's on me, could be done in parallel with 2).
4. Cut the last v2.x.x version just for the sake of it.
5. Cut v3 using maybe separate directory and apply modified version of this diff or at least part of it. Here is where we need to discuss a bit more and I need to learn a bit more. I kinda annoyed that all that go module stuff leaks into the way code itself is organized. But we need to do what we need to do on the other hand.

**Comment by pin** [comment]
Please let me know if plan above is BS. My goal is to
1. Fix wired test issues and connection leaks which probably affect people.
2. Make this thing module-compatible.
3. Make code better. Love the Pack/Unpack idea + I think we need to explicitly code FSM for sender and receiver and extract that FSM to the separate .go file proudly :P

**Comment by sheerun** [comment]
it's not my code ;(

File: packet.go (hunk 1 above)
Comment (sheerun) [reply packet.go:13]: it's not my code ;(

**Comment by sheerun** [comment]
All I wanted to do with this PR is to make this package easily installable for everyone. It should be backward-compatible

Your ideas are great, but it's a lot of work if you want to do it :)

File: packet.go (hunk 1 above)
Comment (pin) [reply packet.go:13]: Oh lol, someone suck as a code reviewer :P

**Comment by pin** [comment]
@VictorLowther, apology for being annoying. Would you like to rebase your #59 ?

I don't mind doing it myself, but I would like to keep the proper attribution: you've spend time working on that fixes and deserve the "blame"... I mean _fame_ :P

Before you actually do that, just let me know, I will rollback #69 so you don't need to rebase twice.

**Comment by pin** [comment]
@VictorLowther, I was too impassioned about your fixes from #59 so rebased them piece by piece myself and landed. Recognition went to PR description...

File: packet.go (hunk 1 above)
Comment (pin) [reply packet.go:13]: Ah, lol, I'm slowpoke :) Some of that is the code from, like, `v1`? I had some wired flashback in the form of blast from the distant past, phew.

**Comment by pin** [comment]
Ok, took some time, but looks like it is good now. Version `v3` is cut and some long-overdue fixes are incorporated into that.

Closing this PR, @sheerun please let me know if you still seeing any issues. Thank you for your work! It helped to understand the modules stuff and also motivated to finally fix bunch of stuff that wanted to be fixed!!

**Comment by sheerun** [comment]
Wonderful! Thank you. I'll let you know if I'll encounter any issues

**Pull request closed without merging**